include/arch/arm/arch/64/mode/machine/debug.h

```diff
@@ -6,4 +6,141 @@
 
 #pragma once
 
+#ifdef ARM_BASE_CP14_SAVE_AND_RESTORE
 
+void aarch64_restore_user_debug_context(tcb_t *target_thread);
+
+/** Determines and carries out what needs to be done for a debug exception.
+ *
+ * This could be handling a single-stepping exception, or a breakpoint or
+ * watchpoint.
+ */
+seL4_Fault_t handleUserLevelDebugException(word_t esr, word_t fault_vaddr);
+bool_t isDebugFault(word_t esr);
+
+#define MAKE_DBGBVR(num) "DBGBVR" #num "_EL1"
+#define MAKE_DBGBCR(num) "DBGBCR" #num "_EL1"
+#define MAKE_DBGWVR(num) "DBGWVR" #num "_EL1"
+#define MAKE_DBGWCR(num) "DBGWCR" #num "_EL1"
```

Comment:
Shouldn't this be `_EL2` when HYP is enabled?

Reply:
I'm pretty sure those registers don't exist.

Reply:
I should have checked. About EL2, the ARM manual says:

[To enable OS debugging, set MDCR_EL2.TDE to 1, which enables the exceptions from EL0 and EL1 to be handled at EL2. This means that the debugger is installed at EL2.](https://developer.arm.com/documentation/102120/0101/Enabling-various-self-hosted-debug-models).

Reply:
Yep, that's done in the `setHDCRTrapDebugExceptionState` function and I've tested that the breakpoint and single-stepping tests still pass when running the kernel with ARM_HYP enabled.